source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,320 @@
+#include "redis_cluster.h"
+
+#include <err.h>
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Redis {
+
+RedisCluster::RedisCluster(
+    const envoy::api::v2::Cluster& cluster,
+    const envoy::config::cluster::redis::RedisClusterConfig& redisCluster,
+    NetworkFilters::Common::Redis::Client::ClientFactory& redis_client_factory,
+    Upstream::ClusterManager& clusterManager, Runtime::Loader& runtime,
+    Network::DnsResolverSharedPtr dns_resolver,
+    Server::Configuration::TransportSocketFactoryContext& factory_context,
+    Stats::ScopePtr&& stats_scope, bool added_via_api)
+    : Upstream::BaseDynamicClusterImpl(cluster, runtime, factory_context, std::move(stats_scope),
+                                       added_via_api),
+      cluster_manager_(clusterManager),
+      cluster_refresh_rate_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(redisCluster, cluster_refresh_rate, 5000))),
+      cluster_refresh_timeout_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(redisCluster, cluster_refresh_timeout, 3000))),
+      dispatcher_(factory_context.dispatcher()), dns_resolver_(std::move(dns_resolver)),
+      dns_lookup_family_(Upstream::getDnsLookupFamilyFromCluster(cluster)),
+      load_assignment_(cluster.has_load_assignment()
+                           ? cluster.load_assignment()
+                           : Config::Utility::translateClusterHosts(cluster.hosts())),
+      local_info_(factory_context.localInfo()), random_(factory_context.random()),
+      redis_discovery_session_(
+          std::make_unique<RedisDiscoverySession>(*this, redis_client_factory)) {
+  const auto& locality_lb_endpoints = load_assignment_.endpoints();
+  for (const auto& locality_lb_endpoint : locality_lb_endpoints) {
+    for (const auto& lb_endpoint : locality_lb_endpoint.lb_endpoints()) {
+      const auto& host = lb_endpoint.endpoint().address();
+      dns_discovery_resolve_targets_.emplace_back(new DnsDiscoveryResolveTarget(
+          *this, host.socket_address().address(), host.socket_address().port_value(),
+          locality_lb_endpoint, lb_endpoint));
+    }
+  }
+};
+
+void RedisCluster::startPreInit() {
+  for (const DnsDiscoveryResolveTargetPtr& target : dns_discovery_resolve_targets_) {
+    target->startResolve();
+  }
+}
+
+void RedisCluster::updateAllHosts(const Upstream::HostVector& hosts_added,
+                                  const Upstream::HostVector& hosts_removed,
+                                  uint32_t current_priority) {
+  Upstream::PriorityStateManager priority_state_manager(*this, local_info_, nullptr);
```

Comment:
I'm a little confused, if this priority state manager is a local variable, what is the purpose of this entire call chain?

Reply:
It updates the priorityState of the cluster.  This is the same as how StrictDns implement it.

Reply:
OK I think I see my confusion. Why does redis_discovery_session_ need to exist? You create it on cluster creation and never recreate it?

Reply:
redis_discovery_session is responsible for interfacing with Redis, like handling the request response.  Initially it was going to handle a single connection, but then it start maintaining a pool of connections, so we never recreate it unless the cluster is recreated.

Reply:
Right, so do we need the class at all? Can't this just be handled directly by the cluster without the object and indirection?

Reply:
I think it makes the structure cleaner, the RedisCluster subclass Upstream::BaseDynamicClusterImpl, and RedisDiscoverySession subclass Redis::Client::Config and Redis::Client::PoolCallbacks.  It's less clean if we do all of those in RedisCluster, and it's got to implement quite a few unrelated interfaces.  Currently the onResponse method on RedisDiscoverySession is not very cleanly implemented, I can refactor that to make it cleaner.

Reply:
OK. I don't feel strongly enough to argue about it. SGTM.

Reply:
RedisCluster handles the BaseDynamicClusterImpl interface and RedisDiscoverySession handles the Redis::Client::Config and Redis::Client::PoolCallbacks.  The current onResponse is not very clean, let me refactor that to make the distinction a bit clearer.